crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs

```diff
@@ -438,8 +434,67 @@ where
         None => farmer_app_info.protocol_info.max_pieces_in_sector,
     };
 
-    let downloading_semaphore = Arc::new(Semaphore::new(sector_downloading_concurrency.get()));
-    let encoding_semaphore = Arc::new(Semaphore::new(sector_encoding_concurrency.get()));
+    let plotting_thread_pool_core_indices =
+        thread_pool_core_indices(plotting_thread_pool_size, sector_encoding_concurrency);
+    let replotting_thread_pool_core_indices = {
+        let mut replotting_thread_pool_core_indices =
+            thread_pool_core_indices(replotting_thread_pool_size, sector_encoding_concurrency);
+        if replotting_thread_pool_size.is_none() {
+            // The default behavior is to use all CPU cores, but for replotting we just want half
+            replotting_thread_pool_core_indices
+                .iter_mut()
+                .for_each(|set| set.truncate(set.cpu_cores().len() / 2));
+        }
+        replotting_thread_pool_core_indices
+    };
+
+    let downloading_semaphore = Arc::new(Semaphore::new(
+        sector_downloading_concurrency
+            .map(|sector_downloading_concurrency| sector_downloading_concurrency.get())
+            .unwrap_or(plotting_thread_pool_core_indices.len() + 1),
+    ));
+
+    let all_cpu_cores = all_cpu_cores();
+    let plotting_thread_pool_manager = create_tokio_thread_pool_manager_for_pinned_nodes(
+        "plotting",
+        plotting_thread_pool_core_indices,
+    )?;
+    let replotting_thread_pool_manager = create_tokio_thread_pool_manager_for_pinned_nodes(
+        "replotting",
+        replotting_thread_pool_core_indices,
+    )?;
+    let farming_thread_pool_size = farming_thread_pool_size
+        .map(|farming_thread_pool_size| farming_thread_pool_size.get())
+        .unwrap_or_else(|| {
+            all_cpu_cores
+                .first()
+                .expect("Not empty according to function description; qed")
+                .cpu_cores()
+                .len()
+        });
+
+    if all_cpu_cores.len() > 1 {
```

Comment:
this is fine but maybe using the topology would give better readability though

Reply:
Can you clarify what you mean here?

Reply:
I meant to say that we can know if there is NUMA cores if the topology is  `Some` and maybe `all_cpu_cores` can have a fn like `is_numa_cores_available` or something so that its much easier to read the code instead of `all_cpu_cores.len() > 1 `

Reply:
`all_cpu_cores` contains CPU cores grouped by NUMA nodes, hence the check. I just wanted to add something to logs that would help users know whether they have NUMA configuration. I'll think about improving this in the future.